### DISCUSSION TITLE
Expose DSHMC.exe's features to the command line for scripting.

### DISCUSSION THREAD

**Discussion by Awebb** [nefarius/DsHidMini#160] [Ideas]
Greetings you fine folks!

I just discovered this tool and it's working great. I think I've checked every resource short of reading the source code and couldn't find a way to script DSHMC.exe for easier "HID device mode" switching. Imagine calling it like this:
```
DSHMC --instanceid _BTHPS3BUS\..._ --hidmode sdf
DSHMC --sdfdigitalbuttons [mute/unmute]
```
The switch layout of course negotiable and I'd be willing to sit down and cook something (reasonable) up.

**_Why would we want this?_**
At this point, switching between PCSX2, RPCS3 and Retroarch requires a couple of clicks. By exposing those options to the CLI, one could launch, say, PCSX2 by script:
```
set HID mode to SDF
start PCSX2
close PCSX2
set HID mode (back) to Xinput or SXS.
```

**Comment by nefarius** [comment]
I understand the need and can definitely get behind this idea. Only big disadvantage I see in adding CLI to DSHMC is the need for administration elevation, that is never a good idea in scripts. Maybe have an IPC mechanism with an elevated process that makes the modification instead, not the actual CLI process, what do you think?

**Reply by Awebb** [reply]
I personally use gsudo to elevate single processes in scripts, but I see how this might not be for everyone. I'm afraid people would then run the whole script elevated, I see the issue.

Perhaps a service that does the changes and the frontend only communicates with it, an IPC mechanism, indeed. This would also eliminate the need to run the GUI as admin.

Different angle: Isn't there a way to elevate "sub processes" within a GUI in C#, that only ask the UAC when you try and run something that needs it?

**Reply by nefarius** [reply]
> Different angle: Isn't there a way to elevate "sub processes" within a GUI in C#, that only ask the UAC when you try and run something that needs it?

There is but IMHO it's very ugly since unlike on *nix systems where there's a syscall here you need to invoke a new process that then runs elevated, I don't like that. Will look into the service idea, I've done similar setups before.
